lib/Feature/XmlParser.cpp

```diff
@@ -0,0 +1,267 @@
+        parseOptions(H, true);
+      } else if (!xmlStrcmp(H->name, reinterpret_cast<constXmlCharPtr>(
+                                         "booleanConstraints"))) {
+        parseConstraints(H);
+      }
+    }
+  }
+}
+
+std::unique_ptr<FeatureModel> XmlParser::buildFeatureModel() {
+  if (!Doc) {
+    return nullptr;
+  }
+  Features.clear();
+  parseVm(xmlDocGetRootElement(Doc.get()));
+  if (Features.find("root") == Features.end()) {
+    Features.try_emplace("root",
+                         std::make_unique<BinaryFeature>("root", false));
+  }
+
+  for (const auto &P : RawEdges) {
+    if (Features.find(P.first) != Features.end() &&
+        Features.find(P.second) != Features.end()) {
+      assert(Features[P.first] && Features[P.second]);
+      Features[P.first]->addChild(Features[P.second].get());
+      Features[P.second]->addParent(Features[P.first].get());
+    }
+  }
+
+  for (const auto &P : RawExcludes) {
+    if (Features.find(P.first) != Features.end() &&
+        Features.find(P.second) != Features.end()) {
+      assert(Features[P.first] && Features[P.second]);
+      Features[P.first]->addExclude(Features[P.second].get());
+    }
+  }
+
+  for (const auto &C : RawConstraints) {
+    Constraints.emplace_back();
+    for (const auto &P : C) {
+      if (Features.find(P.first) != Features.end()) {
+        assert(Features[P.first]);
+        Constraints.back().emplace_back(Features[P.first].get(), P.second);
+      }
+    }
+  }
+
+  for (const auto &C : Constraints) {
+    if (C.size() == 2) {
+      if (C[0].second != C[1].second) {
+        if (C[0].second) { // A || !B
+          C[1].first->addImplication(C[0].first);
+        } else { // !A || B
+          C[0].first->addImplication(C[1].first);
+        }
+      } else if (!(C[0].second || C[1].second)) { // !A || !B
+        C[0].first->addExclude(C[1].first);
+        C[1].first->addExclude(C[0].first);
+      } else if (C[0].second && C[1].second) { // A || B
+        C[0].first->addAlternative(C[1].first);
+        C[1].first->addAlternative(C[0].first);
+      }
+    } else if (C.size() > 2) {
+      bool B = true;
+      for (const auto &P : C) {
+        B &= P.second;
+      }
+      if (B) {
+        for (const auto &P : C) {
+          for (const auto &PP : C) {
+            if (P.first != PP.first) {
+              P.first->addAlternative(PP.first);
+            }
+          }
+        }
+      }
+    }
+  }
+
+  FMFeature *Root = Features["root"].get();
+  for (const auto &P : Features) {
+    if (P.second->isRoot() && P.second.get() != Root) {
+      Root->addChild(P.second.get());
+      P.second->addParent(Root);
+    }
+    for (auto &C : *P.second) {
+      if (C->isOptional()) {
+        P.second->addRelationship(std::make_unique<Optional<FMFeature>>(C));
+      } else {
+        P.second->addRelationship(std::make_unique<Mandatory<FMFeature>>(C));
+      }
+    }
+    // TODO relationships: or and xor
```

Comment:
Should this be fixed? If you plan to fix it later please create an issue and write the TODO like this:
`TODO (se-passau/VaRA#42): do stuff to bar`

Otherwise, I'll always lose track of the TODOs :grinning: